### PR TITLE
DYN Dns Authenticator

### PR DESCRIPTION
**Pull request by trinopoty**
Implement an Authenticator which can fulfill a dns-01 challenge using the DYN API.

Testing Done:
- tox -e py27
- tox -e lint
- Manual Testing
  - Used certbot certonly --dns-dyn -d, specifying a credentials file as a command line argument. Verified that a certificate was successfully obtained without user interaction.
  - Used certbot certonly --dns-dyn -d, without specifying a credentials file as a command line argument. Verified that the user was prompted and that a certificate was successfully obtained.
  - Used certbot certonly -d. Verified that the user was prompted for a credentials file after selecting DYN interactively and that a certificate was successfully obtained.
  - Used certbot renew --force-renewal. Verified that certificates were renewed without user interaction.

**NOTE** This is for the DYN "Managed DNS Service".

### CODE REVIEW

**Comment by schoen** [comment]
@bmw, could you take a look at this also?

**Comment by bmw** [comment]
Does DYN DNS work with our current `certbot-dns-rfc2136` plugin? It looks like [it might](https://dyn.com/updater/tsig/).

**Comment by trinopoty** [comment]
There seems to be some complications regarding this. I'm currently contacting DYN team for more information regarding this.

**Comment by trinopoty** [comment]
I received some information from DYN support. Turns out, DYN has 2 different service for primary DNS, "Managed DNS" and "Standard DNS". Only records in "Standard DNS" service can be updated using the `certbot-dns-rfc2136` plugin since it supports TSIG based updated. The "Managed DNS" service does not support TSIG based updates and can only be updated by the HTTP api.
There's also a "DynDNS Pro" service that supports the TSIG based update system.

The difference between the two is that the "Managed DNS" has a much larger and more redundant network than the "Standard DNS" service.

This plugin targets the "Managed DNS" service.

**Comment by bmw** [comment]
Thanks for digging into that @trinopoty. I'll try and make sure this PR gets reviewed soon.

File: certbot-dns-dyn/certbot_dns_dyn/dns_dyn_test.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    def test_get_dyn_client(self, patch_DynectSession):
+        self.auth._get_dyn_client()
+        patch_DynectSession.assert_called_once_with(DYN_CUSTOMER,
+                                                    DYN_USERNAME,
+                                                    DYN_PASSWORD)
+
+    @mock.patch('certbot_dns_dyn.dns_dyn.get_all_zones')
+    def test_find_zone_missing(self, patch_get_all_zones):
+        patch_get_all_zones.return_value = []
+        assert self.auth._find_zone('_acme-challenge.letsencrypt.org') == None
```
Comment (joohoi) [inline certbot-dns-dyn/certbot_dns_dyn/dns_dyn_test.py]: I think it'd be wise to use `example.com` domain throughout the tests in general instead.

**Comment by trinopoty** [comment]
Any update on the review of this?

**Comment by bmw** [comment]
Was this closed in favor of https://github.com/certbot/certbot/pull/7494? It doesn't look like Lexicon supports Dyn (yet): https://github.com/AnalogJ/lexicon#providers

**Comment by trinopoty** [comment]
No. I closed this mainly because the code will need a lot of changes to bring it up to date with master and given the current DNS plugins situation, it doesn't make sense to put effort into it.
I'll try and reopen this when the DNS plugins situation is resolved.

**Comment by bmw** [comment]
OK. That makes sense.

The DNS plugin issue is still on our radar, but we had other projects we had to commit to this year. I'm hoping we'll have some time to work on this in the next 6 months or so and depending on our solution, DYN is a provider we'd be interested in supporting because of how popular it is.

**Pull request closed without merging**